### ISSUE TITLE
Samtools --write-index option to choose BAI or CSI index

### ISSUE THREAD

**Issue by calliza** [samtools/samtools#1196]
#### Is your feature request related to a problem? Please specify.
--write-index in samtools v1.10 creates CSI indexes. I don't see anything in the documentation to specify what kind of index to output for the BAM. 
#### Describe the solution you would like.
It would be great if samtools --write-index had an option to specify BAI or CSI index output, like the 'index' command does.
Thanks,
Cali

**Comment by jkbonfield** [comment]
On investigation, the answer to this is that it's already possible.

Commit https://github.com/samtools/samtools/pull/1062/commits/004232c563fbd6bdb9c692b2cd9661c9fdbfdea0 added the ability to use the `##idx##` joint filename + index_file notation to specify which file to write the index to, which in turn also means which format.

Eg:

```
samtools view --write-index in.bam -o out.bam##idx##out.bam.bai
```

It appears when this new feature was added, the documentation wasn't updated so I'll keep this issue open until it's documented properly.

**Comment by calliza** [comment]
That's great news, thank you :-)

**Comment by isinaltinkaya** [comment]
Is it possible to use this also with samtools merge since it does not have the -o option?

**Comment by valeriuo** [comment]
Yes, `merge` still requires a name for the output file, to which the index file name can be appended (`alignment.bam##idx##alignment.bam.bai`). Using CLI options (`-o`) just allows the user to specify the arguments in any order. Without them, the arguments are position dependent.

**Comment by outpaddling** [comment]
Might be good to update the samtools-index man page to explain the difference between bai and csi, especially when bai is preferable, since it's the default, but cannot always be used.

**Comment by stevekm** [comment]
considering that `#` is a comment character in some programming languages, and in shell, that seems like an odd choice for this purpose. Can we  just have a dedicated command line argument for this?

**Comment by jkbonfield** [comment]
It was chosen rather specifically because it's a shell meta-character and so (wisely) rarely comes up in real filenames.  The rather cryptic way of specifying two files as a compound single file was done as a generic way of retrofitting this extra information to all sub-commands without needing to add long options for every one, especially given some commands take multiple filenames.  (Also quoting is the solution to shell meta-characters) 

Consider also that this needs to work also on files that read BAM (etc) files too, where maybe the BAM and the associated BAI filenames are some block-store sha hash or similarly opaque names that cannot be automatically derived from the software by simply adding suffixes.

In short, we didn't want to start from here, but retrofitting this afterwards rather constrained the options available.
